store/src/java/com/zimbra/cs/html/OwaspHtmlSanitizer.java

```diff
@@ -1,36 +1,409 @@
+
+        public String apply(String elementName, String attributeName, String value) {
+            value = removeAnySpacesAndEncodedChars(value);
+            value = AV_JS_ENTITY.matcher(value).replaceAll("JS-ENTITY-BLOCKED");
+            value = AV_SCRIPT_TAG.matcher(value).replaceAll("SCRIPT-TAG-BLOCKED");
+            return value;
+        }
+    };
+
+    private AttributePolicy BLOCK_FORM_SAME_HOST_POST_REQ = new AttributePolicy() {
+
+        public String apply(String elementName, String attributeName, String value) {
+            if (sameHostFormPostCheck == true && reqVirtualHost != null) {
+                try {
+                    URL url = new URL(value);
+                    String formActionHost = url.getHost().toLowerCase();
+
+                    if (formActionHost.equalsIgnoreCase(reqVirtualHost)) {
+                        value = value.replace(formActionHost, "SAMEHOSTFORMPOST-BLOCKED");
+                    }
+                } catch (MalformedURLException e) {
+                    ZimbraLog.soap
+                        .warn("Failure while trying to block mailicious code. Check for URL "
+                            + " match between the host and the action URL of a FORM."
+                            + "Error parsing URL, possible relative URL." + e.getMessage());
+                }
+            }
+            return value;
+        }
+    };
+
+    private PolicyFactory POLICY_DEFINITION;
+
+    /**
+     * This sanitizer policy removes these elements and the content within:
+     * <ul>
+     * <li>APPLET</li>
+     * <li>FRAMESET</li>
+     * <li>OBJECT</li>
+     * <li>SCRIPT</li>
+     * <li>STYLE</li>
+     * <li>TITLE</li>
+     * </ul>
+     *
+     * This sanitizer policy removes these elements but preserves the content
+     * within:
+     * <ul>
+     * <li>BASEFONT</li>
+     * <li>FRAME</li>
+     * <li>HEAD</li>
+     * <li>IFRAME</li>
+     * <li>ISINDEX</li>
+     * <li>LINK</li>
+     * <li>META</li>
+     * <li>NOFRAMES</li>
+     * <li>PARAM</li>
+     * <li>NOSCRIPT</li>
+     * </ul>
+     *
+     * This sanitizer policy removes these attributes from all elements:
+     * <ul>
+     * <li>code</li>
+     * <li>codebase</li>
+     * <li>id</li>
+     * <li>for</li>
+     * <li>headers</li>
+     * <li>onblur</li>
+     * <li>onchange</li>
+     * <li>onclick</li>
+     * <li>ondblclick</li>
+     * <li>onfocus</li>
+     * <li>onkeydown</li>
+     * <li>onkeypress</li>
+     * <li>onkeyup</li>
+     * <li>onload</li>
+     * <li>onmousedown</li>
+     * <li>onmousemove</li>
+     * <li>onmouseout</li>
+     * <li>onmouseover</li>
+     * <li>onmouseup</li>
+     * <li>onreset</li>
+     * <li>onselect</li>
+     * <li>onsubmit</li>
+     * <li>onunload</li>
+     * <li>tabindex</li>
+     * </ul>
+     */
+
+    private void instantiatePolicy() {
+        POLICY_DEFINITION = new HtmlPolicyBuilder().allowAttributes("dir")
```

Comment:
note: I'd like to do a side by side comparison with https://github.com/cure53/DOMPurify but that will take a bit more time than what I have at the moment.  We should put together that comparison soon though.

Reply:
Sure.